### PR TITLE
[Fix] Ensure ln_out is not cached if wgrads not required

### PR DESCRIPTION
**Pull request by AllenFarcas**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes https://github.com/ROCm/frameworks-internal/issues/13370

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:
- Fixed logic for not having ln_out saved if weight gradients are not required
- Added unit test to check for layernorm_linear and layernorm_mlp saved tensors logic

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/triton_kernels/test_layernorm_saved_tensors_logic.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
+import pytest
+import torch
+import torch.nn as nn
+from unittest.mock import patch
+from transformer_engine.pytorch import LayerNormLinear, LayerNormMLP, fp8_autocast
+from transformer_engine.pytorch.tensor import QuantizedTensor
+
+@pytest.fixture(scope="module", autouse=True)
+def _check_cuda():
+    if not torch.cuda.is_available() or not torch.cuda.is_bf16_supported():
```
Comment (wenchenvincent) [inline tests/pytorch/triton_kernels/test_layernorm_saved_tensors_logic.py]: We need to check if fp8 support is available or not.

File: tests/pytorch/triton_kernels/test_layernorm_saved_tensors_logic.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            out, ln_out_returned = model(inp)
+            out.backward(grad_output)
+
+    assert mock_backward.called, "The backward method was not called for the `Gradients OFF` test."
+    assert len(saved_ln_out_container) == 1
+    # Assert that `None` was saved
+    assert saved_ln_out_container[0] is None, f"`None` should be saved for `ln_out` when grads are OFF, got {type(saved_ln_out_container[0])} instead."
+
+
+def test_layernorm_linear_saved_tensors_logic():
```
Comment (wenchenvincent) [inline tests/pytorch/triton_kernels/test_layernorm_saved_tensors_logic.py]: These two tests are structurally similar. Can we refactor them into one parameterized test?

File: tests/pytorch/triton_kernels/test_layernorm_saved_tensors_logic.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
+import pytest
```
Comment (wenchenvincent) [inline tests/pytorch/triton_kernels/test_layernorm_saved_tensors_logic.py]: Let's move the file location to `tests/pytorch`

**Review by wenchenvincent — CHANGES_REQUESTED** [review]
Please address the comments. Thanks!

File: transformer_engine/pytorch/module/layernorm_linear.py [hunk 4]
```diff
@@ -353,8 +353,8 @@ def forward(
 
         if not weight.requires_grad:
             if not return_layernorm_output:
-                ln_out = ln_out_total = None
                 clear_tensor_data(ln_out, ln_out_total)
+            ln_out = None
```
Comment (ipanfilo) [inline transformer_engine/pytorch/module/layernorm_linear.py:357]: Later ln_out usage is guarded with if weight.requires_grad else None. So, is the only effective change here removing line 356?

File: tests/pytorch/test_layernorm_saved_tensors_logic.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
+import pytest
```
Comment (ipanfilo) [inline tests/pytorch/test_layernorm_saved_tensors_logic.py:4]: Copyright

**Review by ipanfilo — DISMISSED** [review]

File: transformer_engine/pytorch/module/layernorm_linear.py (hunk 4 above)
Comment (AllenFarcas) [reply transformer_engine/pytorch/module/layernorm_linear.py:357]: This follows what was already present in layernorm_mlp.py
The previous version of line 356 was wrong since they did assignment to None and then try to clear a None's data which doesn't make sense.

Indeed it is redundant with that check but wasn't sure if it wouldn't be better to leave both for safety.

**Review by wenchenvincent — APPROVED** [review]
LGTM.
